lib/ProductOpener/DataQualityFood.pm

```diff
@@ -591,22 +591,59 @@ sub check_nutrition_data_energy_computation ($product_ref) {
 				$computed_energy += $grams * $energy_per_gram;
 			}
 
-			# Compare to specified energy value with a tolerance of 30% + an additiontal tolerance of 5
-			if (   ($computed_energy < ($specified_energy * 0.7 - 5))
-				or ($computed_energy > ($specified_energy * 1.3 + 5)))
-			{
-				# we have a quality problem
-				push @{$product_ref->{data_quality_errors_tags}},
-					"en:energy-value-in-$unit-does-not-match-value-computed-from-other-nutrients";
+			# following error/warning should be ignored for some categories
+			# for example, lemon juices containing organic acid, it is forbidden to display organic acid in nutrition tables but
+			# organic acid contributes to the total energy calculation
+			my $skip_energy_error = 0;
+			if (defined $product_ref->{categories_tags}) {
+				# start from end, categories_tags has the most specific categories at the end
+				my $i = @{$product_ref->{categories_tags}} - 1;
+				while (
+					($i >= 0)
```

Comment:
It seems to me you could rewrite this with a foreach and the reverse function + a break if you match condition to skip energy error. It would be more consistent with the rest of the codebase.

Reply:
Also, you make two call to get_inherited_property while it's a costly method, you'd better store the result in a local variable.

Reply:
yes, thank you for the review and comments. Sorry, I am still learning. I will try to improve it. 

I was inspired by the following snippet in **ProducersFood.pm**
```

	my $i = @{$product_ref->{categories_tags}} - 1;

	while (
		($i >= 0)
		and not((defined $categories_nutriments_ref->{$product_ref->{categories_tags}[$i]})
			and (defined $categories_nutriments_ref->{$product_ref->{categories_tags}[$i]}{nutriments}))
		)
	{
		$i--;
	}
	# categories_tags has the most specific categories at the end

	if ($i >= 0) {
```

Now I reworked it based on **compute_ecoscore_agribalyse** found in **Ecoscore.pm**

Can you review this PR as well: https://github.com/openfoodfacts/openfoodfacts-server/pull/8360? 
I did code something similar there too.